packages/protocol/contracts/common/linkedlists/SortedLinkedListWithMedian.sol

```diff
@@ -0,0 +1,263 @@
+pragma solidity ^0.5.8;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "./LinkedList.sol";
+import "./SortedLinkedList.sol";
+
+
+/**
+ * @title Maintains a sorted list of unsigned ints keyed by bytes32.
+ */
+library SortedLinkedListWithMedian {
+
+  using SafeMath for uint256;
+  using SortedLinkedList for SortedLinkedList.List;
+
+  enum MedianAction {
+    None,
+    Lesser,
+    Greater
+  }
+
+  enum MedianRelation {
+    Undefined,
+    Lesser,
+    Greater,
+    Equal
+  }
+
+  struct List {
+    SortedLinkedList.List list;
+    bytes32 median;
+    mapping(bytes32 => MedianRelation) relation;
+  }
+
+  /**
+   * @notice Inserts an element into a doubly linked list.
+   * @param key The key of the element to insert.
+   * @param value The element value.
+   * @param lesserKey The key of the element less than the element to insert.
+   * @param greaterKey The key of the element greater than the element to insert.
+   */
+  function insert(
+    List storage list,
+    bytes32 key,
+    uint256 value,
+    bytes32 lesserKey,
+    bytes32 greaterKey
+  )
+    public
+  {
+    list.list.insert(key, value, lesserKey, greaterKey);
+    LinkedList.Element storage element = list.list.list.elements[key];
+
+    MedianAction action = MedianAction.None;
+    if (list.list.list.numElements == 1) {
+      list.median = key;
+      list.relation[key] = MedianRelation.Equal;
+    } else if (list.list.list.numElements % 2 == 1) {
+      // When we have an odd number of elements, and the element that we inserted is less than
+      // the previous median, we need to slide the median down one element, since we had previously
+      // selected the greater of the two middle elements.
+      if (element.previousKey == bytes32(0) || list.relation[element.previousKey] == MedianRelation.Lesser) {
+        action = MedianAction.Lesser;
+        list.relation[key] = MedianRelation.Lesser;
+      } else {
+        list.relation[key] = MedianRelation.Greater;
+      }
+    } else {
+      // When we have an even number of elements, and the element that we inserted is greater than
+      // the previous median, we need to slide the median up one element, since we always select
+      // the greater of the two middle elements.
+      if (element.nextKey == bytes32(0) || list.relation[element.nextKey] == MedianRelation.Greater) {
+        action = MedianAction.Greater;
+        list.relation[key] = MedianRelation.Greater;
+      } else {
+        list.relation[key] = MedianRelation.Lesser;
+      }
+    }
+    updateMedian(list, action);
+  }
+
+  /**
+   * @notice Removes an element from the doubly linked list.
+   * @param key The key of the element to remove.
+   */
+  function remove(List storage list, bytes32 key) public {
+    MedianAction action = MedianAction.None;
+    if (list.list.list.numElements == 0) {
```

Comment:
Why does it make sense to zero the median on a `remove()` when the list is empty?
Should control flow continue to `updateMedian(emptyList, none)` and `emptyList.remove(key)`?

Reply:
I think we just want some way of detecting that there is no median, and setting it to the zero key, which is invalid, is the only way I can think of to do that.

This line was supposed to check if `numElements == 1`, oops!

Reply:
I think `numElements == 1` => `median = 0` is general enough. We should require at the top of this function that `numElements > 0` though.